### PR TITLE
ft # 187300245 user should be able to update their passowrd

### PR DESCRIPTION
**Pull request by k3lly003**
### What does this PR do?
This feature implement the functionality that allow the users to update their  password.

### Description of Task to be completed?

The user might want to change  the password in case they want to. this is different from resetting the password they have forgot.
Then users new password should be hashed and updated, or rejected due to the user not remembering old password or new password and confirm password fields not matching. 

### How should this be manually tested?

To test this you should have the swagger documentation link:  https://champs-ec-be.onrender.com/api/docs/

### What are the relevant pivotal tracker/Trello stories?

#187300245

### CODE REVIEW

File: src/database/seeders/20240325212203-user.js [hunk 1]
```diff
@@ -1,77 +1,81 @@
-import { v4 as uuidv4 } from "uuid";
+/* eslint-disable @typescript-eslint/no-var-requires */
+const { v4: uuidv4 } = require("uuid");
```
[the remark this answers is not in the document]
Comment (emmanueltct) [reply src/database/seeders/20240325212203-user.js:2]: - use the passwordEncrypt function to encrypt user's new password 
- passwordCompare function to compare user password and old password 
both functions are created in an encrypts file just import them

[the remark this answers is not in the document]
Comment (emmanueltct) [reply src/database/seeders/20240325212203-user.js:2]: use validation middleware to validate user input filed

**Review by emmanueltct — COMMENTED** [review]
- use the passwordEncrypt function to encrypt user's new password 
- passwordCompare function to compare user password and old password 
both functions are created in an encrypts file just import  and use them

**Comment by emmanueltct** [comment]
use validation middleware to validate user input fileds

File: src/controllers/user.controllers.ts
Comment (emmanueltct) [inline src/controllers/user.controllers.ts:1]: use validation middleware to check user  input fields are not empty and meet other password conditions
 use created  functions in utils to hash and compare the password

**Review by emmanueltct — CHANGES_REQUESTED** [review]
- use the passwordEncrypt function to encrypt user's new password 
- passwordCompare function to compare user password and old password 
   both functions are created in an encrypts file just import  and use them
-use validation middleware to check the user inputs

**Comment by k3lly003** [comment]
> * use the passwordEncrypt function to encrypt user's new password
> * passwordCompare function to compare user password and old password
>   both functions are created in an encrypts file just import  and use them
>   -use validation middleware to check the user inputs

Issues Resolved ✅

**Review by emmanueltct — APPROVED** [review]

**Review by ErnestTchami — APPROVED** [review]

File: src/controllers/user.controllers.ts [hunk 2]
```diff
@@ -172,3 +172,44 @@ export const editUser = async (req: any, res: Response) => {
...
+    const passwordMatch = await passwordCompare(
+      user?.dataValues.password,
+      oldPassword
+    );
+
+    if (!passwordMatch) {
+      return res.status(400).send({ error: "Incorrect old password" });
+    }
+    if (!newPassword) {
+      return res.status(400).send({ error: "New password is required" });
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: This should be done in validation middleware

File: src/controllers/user.controllers.ts [hunk 3]
```diff
@@ -172,3 +172,44 @@ export const editUser = async (req: any, res: Response) => {
...
+  try {
+    const logedInUser: any = req.user;
+    const { oldPassword, newPassword, confirmPassword } = req.body;
+    const user: User | any = await User.findOne({
+      where: {
+        id: logedInUser.dataValues.id
+      }
+    });
+    if (!oldPassword) {
+      return res.status(400).send({
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: This should be done in validation

File: src/controllers/user.controllers.ts [hunk 4]
```diff
@@ -172,3 +172,44 @@ export const editUser = async (req: any, res: Response) => {
...
+        error: "old password are required"
+      });
+    }
+    const passwordMatch = await passwordCompare(
+      user?.dataValues.password,
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: This should be done in validation

File: src/controllers/user.controllers.ts [hunk 5]
```diff
@@ -172,3 +172,44 @@ export const editUser = async (req: any, res: Response) => {
     res.status(400).json({ error: "there is an error user are not updated" });
   }
 };
+export const updateUserPassword = async (req: Request, res: Response) => {
+  try {
+    const logedInUser: any = req.user;
```
Comment (Eli250) [inline src/controllers/user.controllers.ts]: Don't use any

File: src/utils/validations/user.validations.ts [hunk 6]
```diff
@@ -11,4 +11,27 @@ const userLoginValidation = z
...
+  newPassword: z
+    .string()
+    .min(8, "password of length less than 8")
+    .regex(
+      passwordStrength,
+      "You are using wrong new password please try again"
+    ),
+  confirmPassword: z
+    .string()
+    .min(8, "password of length less than 8")
```
Comment (Eli250) [inline src/utils/validations/user.validations.ts]: Here you can instead check if this password equals the new password provided.
You can refer to this [example](https://stackoverflow.com/questions/75572170/zod-validation-based-on-another-field) or [this](https://stackoverflow.com/questions/75572170/zod-validation-based-on-another-field).

**Review by Eli250 — CHANGES_REQUESTED** [review]
Address given feedback

**Review by Eli250 — APPROVED** [review]

**Review by ErnestTchami — APPROVED** [review]

**Pull request merged**